### PR TITLE
Support Managed Identity OAuth2 authentication for Spring ecosystem of kafka in Spring Cloud Azure

### LINKED ISSUE #28708: Library enabling Managed Identity OAuth for spring kafka sub-projects

### PR DESCRIPTION
**Pull request by yiliuTo**
# Description

Resolve #28708

Support Managed Identity based OAuth2 authentication for spring kafka associated projects.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/AzureEventHubsKafkaAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * Provide Managed Identity-based OAuth2 authentication for Event Hubs for Kafka.
+ *
+ * @since 4.3.0
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(KafkaTemplate.class)
+@AutoConfigureBefore(KafkaAutoConfiguration.class)
+@ConditionalOnProperty(prefix = AzureEventHubsKafkaProperties.PREFIX, value = "enabled", havingValue = "true",
+        matchIfMissing = true)
+@EnableConfigurationProperties({AzureEventHubsKafkaProperties.class, KafkaProperties.class})
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/AzureEventHubsKafkaAutoConfiguration.java:30]: Do we need the `KafkaProperties` here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/AzureEventHubsKafkaAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+public class AzureEventHubsKafkaAutoConfiguration {
+
+    public AzureEventHubsKafkaAutoConfiguration(AzureGlobalProperties azureGlobalProperties,
+                                                AzureEventHubsKafkaProperties azureEventHubsKafkaProperties) {
+        AzurePropertiesUtils.copyPropertiesIgnoreTargetNonNull(azureGlobalProperties.getProfile(),
+                azureEventHubsKafkaProperties.getProfile());
+        AzurePropertiesUtils.copyPropertiesIgnoreTargetNonNull(azureGlobalProperties.getCredential(),
+                azureEventHubsKafkaProperties.getCredential());
+    }
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/AzureEventHubsKafkaAutoConfiguration.java:39]: We'd better return a `AzureEventHubsKafkaProperties` with the default value coming from `AzureGlobalProperties` as other autoconfigurations do.

File: sdk/spring/CHANGELOG.md [hunk 3]
```diff
@@ -3,8 +3,10 @@
 ## 4.3.0-beta.1 (Unreleased)
 
 ### Features Added
+- Support OAuth2 authentication for Spring ecosystems of Kafka [#29135](https://github.com/Azure/azure-sdk-for-java/pull/29135).
 
 ### Breaking Changes
+- Drop support of connection string or Azure Resource Manager based authentication for Spring ecosystems of Kafka [#29135](https://github.com/Azure/azure-sdk-for-java/pull/29135).
```
Comment (saragluna) [inline sdk/spring/CHANGELOG.md:9]: we should follow the same style in the changelog as earlier releases.

File: sdk/spring/spring-cloud-azure-service/src/main/java/com/azure/spring/cloud/service/implementation/kafka/AzureOAuthBearerToken.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * Implementation of {@link OAuthBearerToken} for Azure Event Hubs.
+ */
+public class AzureOAuthBearerToken implements OAuthBearerToken {
+    private final String token;
+    private final Long startTimeMs;
+    private final long lifetimeMs;
+    private final Set<String> scope;
+    private final String principalName;
+
+    public AzureOAuthBearerToken(String token, Long startTimeMs, long lifetimeMs, Set<String> scope,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-service/src/main/java/com/azure/spring/cloud/service/implementation/kafka/AzureOAuthBearerToken.java:20]: Should this class accept an Azure core AccessToken?

File: sdk/spring/spring-cloud-azure-service/src/test/java/com/azure/spring/cloud/service/kafka/KafkaOAuth2AuthenticateCallbackHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import com.azure.core.credential.AccessToken;
+import com.azure.spring.cloud.service.implementation.kafka.AzureOAuthBearerToken;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-service/src/test/java/com/azure/spring/cloud/service/kafka/KafkaOAuth2AuthenticateCallbackHandlerTest.java:14]: We'd add more test cases for the EventHubsAutoconfiguration

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/properties/utils/AzureEventHubsKafkaPropertiesUtils.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Utils class for {@link KafkaProperties} String constants.
+ */
+public final class AzureEventHubsKafkaPropertiesUtils {
+
+    private AzureEventHubsKafkaPropertiesUtils() {
+
+    }
+
+    public static final String SECURITY_PROTOCOL_CONFIG_SASL = "SASL_SSL";
```
Comment (jialigit) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/properties/utils/AzureEventHubsKafkaPropertiesUtils.java:16]: how about consider using the original const in `kafka-client` lib.
for example:
SecurityProtocol.SASL_SSL 
OAuthBearerLoginModule.OAUTHBEARER_MECHANISM

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/KafkaPropertiesBeanPostProcessor.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /**
+     * Detect whether we need to configure SASL/OAUTHBEARER properties for {@link KafkaProperties}. Will configure
+     * when the security protocol is not configured, or it's set as SASL_SSL with sasl mechanism as null or OAUTHBEAR.
+     *
+     * @param sourceProperties the source kafka properties for admin/consumer/producer to detect
+     * @return whether we need to configure with Spring Cloud Azure MSI support or not.
+     */
+    private boolean ifSaslOAuthNeedConfigure(Map<String, Object> sourceProperties) {
```
Comment (jialigit) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/KafkaPropertiesBeanPostProcessor.java:75]: how about "needConfigureSaslOAuth"?

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/util/AzurePropertiesUtils.java [hunk 8]
```diff
@@ -99,10 +99,20 @@ public static <T extends AzureProperties> void mergeAzureCommonProperties(AzureP
...
         BeanUtils.copyProperties(source, target, findNullPropertyNames(source));
     }
 
+    /**
+     * Copy properties from source object to target object. Ignore the source value if the related target is non-null.
+     *
+     * @param source The source object.
+     * @param target The target object.
+     */
+    public static void copyPropertiesIgnoreTargetNonNull(Object source, Object target) {
```
Comment (jialigit) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/util/AzurePropertiesUtils.java:112]: Why not directly use:
BeanUtils.copyProperties(source, target, findPropertyName(xxx));

seems it not add meaning or reduce some codes?

File: sdk/spring/spring-cloud-azure-service/src/main/java/com/azure/spring/cloud/service/implementation/kafka/AzureOAuthBearerToken.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import java.time.Instant;
+import java.util.Set;
+
+/**
+ * Implementation of {@link OAuthBearerToken} for Azure Event Hubs.
+ */
+public class AzureOAuthBearerToken implements OAuthBearerToken {
+    private final String token;
+    private final Long startTimeMs;
+    private final long lifetimeMs;
```
[the remark this answers is not in the document]
Comment (jialigit) [reply sdk/spring/spring-cloud-azure-service/src/main/java/com/azure/spring/cloud/service/implementation/kafka/AzureOAuthBearerToken.java:16]: lifetimeMs -> lifeTimeMs?

**Comment by yiliuTo** [comment]
Refactor and migrate to pr #29404

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/properties/utils/AzureEventHubsKafkaPropertiesUtils.java (hunk 6 above)
Comment (yiliuTo) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/properties/utils/AzureEventHubsKafkaPropertiesUtils.java:16]: good catch, updated in #29404

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/KafkaPropertiesBeanPostProcessor.java (hunk 7 above)
Comment (yiliuTo) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/KafkaPropertiesBeanPostProcessor.java:75]: updated in https://github.com/Azure/azure-sdk-for-java/pull/29404

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/AzureEventHubsKafkaAutoConfiguration.java (hunk 2 above)
Comment (yiliuTo) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/kafka/AzureEventHubsKafkaAutoConfiguration.java:39]: Resolved in #29404

File: sdk/spring/CHANGELOG.md (hunk 3 above)
Comment (yiliuTo) [reply sdk/spring/CHANGELOG.md:9]: Resolved in #29404

File: sdk/spring/spring-cloud-azure-service/src/main/java/com/azure/spring/cloud/service/implementation/kafka/AzureOAuthBearerToken.java (hunk 4 above)
Comment (yiliuTo) [reply sdk/spring/spring-cloud-azure-service/src/main/java/com/azure/spring/cloud/service/implementation/kafka/AzureOAuthBearerToken.java:20]: Resolved in #29404

**Pull request closed without merging**